File: 3. Behavioral Design Patterns/State/Javascript/StateExample.js
// State Pattern Example in JavaScript
// TrafficLight cycles through Red, Yellow, Green states

class TrafficLightState { next(context){} getColor(){} }

class TrafficLight {
  constructor(){ this.state = new RedState() }
  setState(s){ this.state = s }
  next(){ this.state.next(this) }
  getColor(){ return this.state.getColor() }
}

class RedState extends TrafficLightState { next(t){ t.setState(new GreenState()) } getColor(){ return 'Red' } }
class YellowState extends TrafficLightState { next(t){ t.setState(new RedState()) } getColor(){ return 'Yellow' } }
class GreenState extends TrafficLightState { next(t){ t.setState(new YellowState()) } getColor(){ return 'Green' } }

;(function main(){
  const light = new TrafficLight()
  for(let i=0;i<6;i++){ console.log(`Light: ${light.getColor()}`); light.next() }
})()

File: 3. Behavioral Design Patterns/State/Kotlin/StateExample.kt
// State Pattern Example in Kotlin
// TrafficLight cycles through Red, Yellow, Green states

interface TrafficLightState { fun next(context: TrafficLight); fun getColor(): String }

class TrafficLight { private var state: TrafficLightState = RedState(); fun setState(s: TrafficLightState){ state = s }; fun next(){ state.next(this) }; fun getColor() = state.getColor() }

class RedState : TrafficLightState { override fun next(t: TrafficLight) = t.setState(GreenState()); override fun getColor() = "Red" }
class YellowState : TrafficLightState { override fun next(t: TrafficLight) = t.setState(RedState()); override fun getColor() = "Yellow" }
class GreenState : TrafficLightState { override fun next(t: TrafficLight) = t.setState(YellowState()); override fun getColor() = "Green" }

fun main(){
    val light = TrafficLight()
    repeat(6){ println("Light: ${light.getColor()}"); light.next() }
}
